### PR TITLE
S4583: Cleanup

### PR DESCRIPTION
**Pull request by martin-strecker-sonarsource**
~see https://github.com/SonarSource/sonar-dotnet/pull/5569#issuecomment-1102688699 why this is still a draft.~

### CODE REVIEW

**Comment by martin-strecker-sonarsource** [comment]
I can't come up with test cases to show how the SyntaxKinds for record and record struct declarations make any difference. The reason is that the analyzer works like this:

1. Find a `BeginInvoke` invocation
1. Walk up the AST until any of the `ParentDeclarationKinds` is found. In most cases, this is any kind of method declaration, but can also be a class/record declaration (e.g. in cases like a field initializer: `public IAsyncResult result = caller.BeginInvoke(name: "Foo", 42, callback: null, @object: null);`)
1. From that point, a visitor is called, which visits every node below, except the visited node is of any of the `ParentDeclarationsKinds`! (This is a bit odd/unexpected behavior IMHO) (see https://github.com/SonarSource/sonar-dotnet/blob/c734b13f17351babe8de3f6b1a3fcfaa88981b9b/analyzers/src/SonarAnalyzer.Common/Rules/BeginInvokePairedWithEndInvokeBase.cs#L110-L112). These nodes are skipped.
1. So for classes/records basically only fields are visited and I can't come up with an `EndInvoke` call on a field initializer because `EndInvoke` returns `void`.

From my point of view, `ParentDeclarationKinds` of type `ClassDeclaration` and the like are just used to limit the scope of the Visitor (3), but the Visitor never finds an `EndInvoke` for any of the `ClassDeclaration`... node types.

**Comment by pavel-mikula-sonarsource** [comment]
There's no reasonable scenario to make this work in a testable way => it doesn't make sense as the rule is currently implemented.

Let's 
* remove everything class-like from the `ParentDeclarationKinds` for C# and VB.NET (class, compilation unit, interface, records, struct) to keep the implementation clean and focused.
* remove the rule from the MMF
* rename this PR to a rule cleanup
* change `ParentMethodCContainsEndInvoke` like this, because it is wrong by-design (can have `null`, but cannot process it and ends in NRE)
```
        private bool ParentMethodContainsEndInvoke(SyntaxNode node, SemanticModel semantic) =>
            node.AncestorsAndSelf().FirstOrDefault(IsParentDeclaration) is { } parentContext
            && IsParentDeclarationWithEndInvoke(parentContext, semantic);
```

**Review by pavel-mikula-sonarsource — CHANGES_REQUESTED** [review]
Let's change the scope to cleanup

**Comment by martin-strecker-sonarsource** [comment]
* [ ] remove everything class-like from the ParentDeclarationKinds for C# and VB.NET (class, compilation unit, interface, records, struct) to keep the implementation clean and focused.
* [ ] remove the rule from the MMF
* [ ] rename this PR to a rule cleanup
* [ ] change ParentMethodCContainsEndInvoke like this, because it is wrong by-design (can have null, but cannot process it and ends in NRE)

File: analyzers/src/SonarAnalyzer.Common/Rules/BeginInvokePairedWithEndInvokeBase.cs [hunk 1]
```diff
@@ -108,8 +106,7 @@ public EndInvokeContext(BeginInvokePairedWithEndInvokeBase<TSyntaxKind, TInvocat
             }
 
             public bool Visit(SyntaxNode node) =>
-                !ContainsEndInvoke  // Stop visiting once we found it
-                && (node == Root || !rule.ParentDeclarationKinds.Contains(rule.Language.Syntax.Kind(node)));
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Rules/BeginInvokePairedWithEndInvokeBase.cs:112]: Peach check should reveal any FP/FN that the removal of this line may cause.

**Review by pavel-mikula-sonarsource — APPROVED** [review]
LGTM

**Pull request merged**